httpbp/client_middlewares.go

```diff
@@ -54,11 +69,43 @@ func NewClient(config ClientConfig, middlewares ...ClientMiddleware) *http.Clien
 // can be returned. If configured the caller needs to call
 // httpbp.DrainAndClose(resp.Body) to ensure the underlying TCP connection can
 // be re-used.
+//
+// Instead of doing this:
+//
+//     // bad example, DO NOT USE
+//     resp, err := client.Do(req)
+//     if err != nil {
+//       // TODO: handle error
+//       return err
+//     }
+//     defer httpbp.DrainAndClose(resp.Body)
+//     // TODO: handle resp
+//
+// Do this:
+//
+//     resp, err := client.Do(req)
+//     if resp != nil {
+//       defer httpbp.DrainAndClose(resp.Body)
+//     }
+//     if err != nil {
+//       // TODO: handle error
+//       return err
+//     }
+//     // TODO: handle resp
+//
 func ClientErrorWrapper() ClientMiddleware {
 	return func(next http.RoundTripper) http.RoundTripper {
-		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
 			resp, err = next.RoundTrip(req)
-			err = ClientErrorFromResponse(resp)
+			if err != nil {
+				return resp, err
+			}
+
+			defer DrainAndClose(resp.Body)
```

Comment:
🔕 I am still unsure why this needs to be done here, and I think it deserves a comment here that explains the necessity.

Reply:
This is definitely wrong. Just tripped over it while writing tests, we only want to drain and close if `ClientErrorFromResponse` returns an error, in that instance it seems to be okay to close since the API user should handle the error first before attempting to read from the body.